sparse/tests/_utils.py

```diff
@@ -0,0 +1,325 @@
+        )
+    )
+    value_shape = draw(
+        array_shapes(min_dims=4, max_dims=6, min_side=3, max_side=5),
+    )
+
+    return shape, index, value_shape
+
+
+@composite
+def gen_transpose(draw):
+    a = draw(st.lists(st.integers(min_value=1, max_value=5), min_size=2, max_size=4))
+    if len(a) is 2:
+        b = draw(st.sampled_from([(1, 0), (0, 1)]))
+    elif len(a) is 3:
+        b = draw(st.sampled_from([(0, 2, 1), (2, 0, 1), (1, 2, 0)]))
+    else:
+        b = draw(st.sampled_from([(0, 3, 2, 1), (1, 0, 3, 2), (3, 2, 0, 1)]))
+
+    s = draw(gen_sparse_random(a, density=0.5, format="gcxs"))
+
+    return s, b
+
+
+@composite
+def gen_sparse_random(draw, shape, **kwargs):
+    seed = draw(st.integers(min_value=0, max_value=100))
+    return sparse.random(shape, random_state=seed, **kwargs)
+
+
+@composite
+def gen_reductions(draw, function=False):
+    reduction = draw(
+        st.sampled_from(
+            ["sum", "mean", "prod", "max", "min", "std", "var", "any", "all"]
+        )
+    )
+    kwargs = {}
+    if reduction not in {"max", "min", "any", "all"}:
+        dtype = draw(st.sampled_from([np.float32, np.float64, None]))
+        kwargs["dtype"] = dtype
+    kwargs["axis"] = draw(st.sampled_from([None, 0, 1, 2, (0, 2), -3, (1, -1)]))
+    kwargs["keepdims"] = draw(st.sampled_from([True, False]))
+    if function:
+        reduction = getattr(np, reduction)
+
+    return reduction, kwargs
+
+
+@composite
+def gen_broadcast_shape(draw):
+    shape1 = draw(
+        st.lists(st.integers(min_value=2, max_value=5), min_size=2, max_size=3).map(
+            tuple
+        )
+    )
+    shape2 = draw(broadcastable_shapes(shape1, min_side=3))
+
+    return shape1, shape2
+
+
+@composite
+def gen_matmul_warning(draw):
+    a = draw(
+        st.sampled_from(
+            [
+                sparse.GCXS.from_numpy(
+                    np.random.choice(
+                        [0, np.nan, 2], size=[100, 100], p=[0.99, 0.001, 0.009]
+                    )
+                ),
+                sparse.COO.from_numpy(
+                    np.random.choice(
+                        [0, np.nan, 2], size=[100, 100], p=[0.99, 0.001, 0.009]
+                    )
+                ),
+                sparse.GCXS.from_numpy(
+                    np.random.choice(
+                        [0, np.nan, 2], size=[100, 100], p=[0.99, 0.001, 0.009]
+                    )
+                ),
+                np.random.choice(
+                    [0, np.nan, 2], size=[100, 100], p=[0.99, 0.001, 0.009]
+                ),
+            ]
+        )
+    )
+    if not isinstance(a, np.ndarray):
+        b = draw(
+            st.sampled_from(
+                [sparse.random((100, 100), density=0.01), scipy.sparse.random(100, 100)]
+            )
+        )
+    else:
+        b = draw(st.sampled_from([sparse.random((100, 100), density=0.01)]))
+
+    return a, b
+
+
+@composite
+def gen_broadcast_shape_dot(draw):
+    a_shape = draw(
```

Comment:
1. Generate ANY two shapes with `1 <= ndim <= 4`.
2. If `ndim_a == 1: shape_b[-1] = shape_a[0]`
3. If `ndim_b == 1: shape_b[0] = shape_a[-2 if ndim_a != 1 else -1]`
4. Else `shape_b[-1] = shape_a[-2]`.

Reply:
1. For `matmul`, do the same as dot but with `ndim` fixed to `1 <= ndim <= 2`.
2. Then generate two separate broadcastable shapes `shape_sa, shape_sb` s.t. both have ndim between 0 and 3. If `ndim_a == 2: shape_a += shape_sa`
4. Repeat step 3 for `b`.